drivers/usdhc/imx/usdhc.c

```diff
@@ -0,0 +1,931 @@
+       > The user data area capacity is calculated from C_SIZE as follows:
+       >
+       >             memory capacity = (C_SIZE+1) * 512KByte
+    */
+    uint64_t c_size = (csd & SD_CSD_C_SIZE_MASK) >> SD_CSD_C_SIZE_SHIFT;
+    blk_config->size = (c_size + 1) * (512 * 1024) / BLK_TRANSFER_SIZE;
+
+    LOG_DRIVER("size: %lu\n", blk_config->size);
+
+    __atomic_store_n(&blk_config->ready, true, __ATOMIC_RELEASE);
+    LOG_DRIVER("Driver initialisation complete\n");
+}
+
+void handle_clients(bool is_irq) {
+    static blk_request_code_t req_code;
+    static uintptr_t req_offset;
+    static uint32_t req_block_number;
+    static uint16_t req_count;
+    static uint32_t req_id;
+    int err;
+
+    // TODO: Handle overflow with this multiplication...?
+    uint32_t block_to_sectors = BLK_TRANSFER_SIZE / SD_BLOCK_SIZE;
+
+    static enum {
+        StateIdle,
+        StateInflight,
+    } state = StateIdle;
+
+    static data_transfer_state_t data_transfer_state = DataStateInit;
+
+    if (!is_irq && state != StateIdle) {
+        /* Only do something with client requests when we're idle */
+        return;
+    }
+
+    switch (state) {
+        case StateIdle:
+            err = blk_dequeue_req(&blk_queue, &req_code, &req_offset, &req_block_number, &req_count, &req_id);
+            if (err == -1) {
+                // no requests to handle
+                return;
+            }
+
+            LOG_DRIVER("Received command: code=%d, offset=0x%lx, block_number=%d, count=%d, id=%d\n",
+                       req_code, req_offset, req_block_number, req_count, req_id);
+
+            state = StateInflight;
+            fallthrough;
+
+        case StateInflight: {
+            blk_response_status_t status;
+            uint16_t success_count = 0;
+            poll_t poll;
+
+            switch (req_code) {
+            case READ_BLOCKS:
+                poll = usdhc_read_blocks(&data_transfer_state, req_offset, req_block_number * block_to_sectors, req_count * block_to_sectors);
+                if (poll == PollPending) return;
+                data_transfer_state = DataStateInit;
+
+                status = SUCCESS;
+                success_count = 1;
+                break;
+
+            case WRITE_BLOCKS:
+                poll = usdhc_write_blocks(&data_transfer_state, req_offset, req_block_number * block_to_sectors, req_count * block_to_sectors);
+                if (poll == PollPending) return;
+                data_transfer_state = DataStateInit;
+
+                status = SUCCESS;
+                success_count = 1;
+                break;
+
+            case FLUSH:
+            case BARRIER:
+                /* No-ops. */
+                status = SUCCESS;
+                success_count = req_count;
+                break;
+
+            default:
+                LOG_DRIVER_ERR("Unknown command code: %d\n", req_code);
+                return;
+            }
+
+            blk_enqueue_resp(&blk_queue, status, success_count, req_id);
```

Comment:
We should always assert that the enqueue works. In this case I believe there should always be space in the virtualiser, so we can assert instead.